### ISSUE TITLE
eventmachine: RuntimeError: rb_fd_init not implemented

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1623]
After changes in https://github.com/eventmachine/eventmachine/pull/878 by @eregon it's possible to compile eventmachine now. Unfortunately, every time when `EM.run` is used, it fails on `RuntimeError: 0x123efff80`
 
To reproduce:

```
git clone https://github.com/eventmachine/eventmachine.git
cd eventmachine
bundle
bundle exec rake compile
ruby -Ilib -e 'require "eventmachine"; EM.run {}'
```

Error:

```
~/.rubies/truffleruby-1.0.0-rc13/lib/truffle/truffle/cext.rb:1130:in `rb_exc_raise': 0x123efff80 (RuntimeError)
	from ruby.c:1971:in `rb_exc_raise'
	from ~/dev/bugs/truffleruby/eventmachine/ext/cmain.cpp:41:in `ensure_eventmachine'
	from ~/dev/bugs/truffleruby/eventmachine/ext/cmain.cpp:71:in `evma_release_library'
	from ~/dev/bugs/truffleruby/eventmachine/ext/rubymain.cpp:884:in `t_release_machine'
	from ~/.rubies/truffleruby-1.0.0-rc13/lib/truffle/truffle/cext_ruby.rb:41:in `release_machine'
	from ~/dev/bugs/truffleruby/eventmachine/lib/eventmachine.rb:204:in `run'
	from -e:1:in `<main>'
```

**Comment by eregon** [comment]
Here is what I get when running on latest master:
```
$ ruby -Ilib -e 'require "eventmachine"; EM.run {}'
lib/truffle/truffle/cext.rb:1163:in `rb_exc_raise': eventmachine not initialized: evma_release_library (RuntimeError)
	from ruby.c:2099:in `rb_exc_raise'
	from /home/eregon/code/eventmachine/ext/cmain.cpp:41:in `ensure_eventmachine'
	from /home/eregon/code/eventmachine/ext/cmain.cpp:71:in `evma_release_library'
	from /home/eregon/code/eventmachine/ext/rubymain.cpp:884:in `t_release_machine'
	from lib/truffle/truffle/cext_ruby.rb:37:in `release_machine'
	from /home/eregon/code/eventmachine/lib/eventmachine.rb:204:in `run'
	from -e:1:in `<main>'
```

So we get a proper error message now at least.
That error seems to mean the global `EventMachine` variable is not set, likely because initialization fails.

Running with `-d` (or the better `--backtraces-raise`) gives some extra details:
```
$ ruby --experimental-options --backtraces-raise -Ilib -e 'require "eventmachine"; EM.run {}'
lib/mri/resolv.rb:170:in `Hosts': LoadError (LoadError)
	from lib/mri/resolv.rb:168:in `Resolv'
	from lib/mri/resolv.rb:38:in `<top (required)>'
	from /home/eregon/code/eventmachine/lib/eventmachine.rb:39:in `require'
	from /home/eregon/code/eventmachine/lib/eventmachine.rb:39:in `require'
	from /home/eregon/code/eventmachine/lib/eventmachine.rb:39:in `<top (required)>'
	from -e:1:in `require'
	from -e:1:in `require'
	from -e:1:in `<main>'
ruby.c:2944:in `rb_tr_error': rb_fd_init not implemented (RuntimeError)
	from ruby.c:3625:in `rb_fd_init'
	from /home/eregon/code/eventmachine/ext/em.cpp:915:in `SelectData_t'
	from /home/eregon/code/eventmachine/ext/em.cpp:155:in `EventMachine_t'
	from /home/eregon/code/eventmachine/ext/cmain.cpp:61:in `evma_initialize_library'
	from /home/eregon/code/eventmachine/ext/rubymain.cpp:233:in `t_initialize_event_machine'
	from lib/truffle/truffle/cext_ruby.rb:37:in `initialize_event_machine'
	from /home/eregon/code/eventmachine/lib/eventmachine.rb:181:in `run'
	from -e:1:in `<main>'
lib/truffle/truffle/cext.rb:1163:in `rb_exc_raise': eventmachine not initialized: evma_release_library (RuntimeError)
	from ruby.c:2099:in `rb_exc_raise'
	from /home/eregon/code/eventmachine/ext/cmain.cpp:41:in `ensure_eventmachine'
	from /home/eregon/code/eventmachine/ext/cmain.cpp:71:in `evma_release_library'
	from /home/eregon/code/eventmachine/ext/rubymain.cpp:884:in `t_release_machine'
	from lib/truffle/truffle/cext_ruby.rb:37:in `release_machine'
	from /home/eregon/code/eventmachine/lib/eventmachine.rb:204:in `run'
	from -e:1:in `<main>'
lib/truffle/truffle/cext.rb:1163:in `rb_exc_raise': eventmachine not initialized: evma_release_library (RuntimeError)
	from ruby.c:2099:in `rb_exc_raise'
	from /home/eregon/code/eventmachine/ext/cmain.cpp:41:in `ensure_eventmachine'
	from /home/eregon/code/eventmachine/ext/cmain.cpp:71:in `evma_release_library'
	from /home/eregon/code/eventmachine/ext/rubymain.cpp:884:in `t_release_machine'
	from lib/truffle/truffle/cext_ruby.rb:37:in `release_machine'
	from /home/eregon/code/eventmachine/lib/eventmachine.rb:204:in `run'
	from -e:1:in `<main>'
```

The LoadError is expected and rescued, so it's fine.
`rb_fd_init not implemented` seems the issue.

**Comment by eregon** [comment]
Note: https://github.com/oracle/truffleruby/issues/1634 is also about `rb_fd_init` not being implemented.

**Comment by eregon** [comment]
I implemented  `rb_fd_init` and friends in 982f68355c8710ef5923b81cab818fba1d228c9b.
That allows to get further for EventMachine and require it + `EM.run{}`, but EventMachine tests don't pass yet (I didn't investigate further).

I'd like to close this issue, because that error no longer happens.
